<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title> Module 3 | SNP calling</title>
    <meta name="author" content="Sophie Colette">
    <meta name="description" content="Module 3: SNP calling - from fastq to vcf">
    <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website">
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04">
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous">
    <link defer rel="stylesheet" href="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.css">
    <link rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5">
    <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons">
    <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?19f3075a2d19613090fe9e16b564e1fe" media="" id="highlight_theme_light">
    <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20
																xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%E2%9A%9B%EF%B8%8F&lt;/text&gt;&lt;/svg&gt;">
    <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e">
    <link rel="canonical" href="https://sc-git1.github.io/blog/2021/distill/">
    <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?e74e74bf055e5729d44a7d031a5ca6a5" media="none" id="highlight_theme_dark">
    <script src="/assets/js/theme.js?96d6b3e1c3604aca8b6134c7afdd5db6"></script>
    <script src="/assets/js/dark_mode.js?9b17307bb950ffa2e34be0227f53558f"></script>
    <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.0.8/dist/medium-zoom.min.js" integrity="sha256-7PhEpEWEW0XXQ0k6kQrPKwuoIomz8R8IYyuU1Qew4P8=" crossorigin="anonymous"></script>
    <script defer src="/assets/js/zoom.js?7b30caa5023af4af8408a472dc4e1ebb"></script>
    <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script>
    <script type="text/javascript">
      window.MathJax = {
        tex: {
          tags: "ams"
        }
      };
    </script>
    <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script>
    <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
    <script src="/assets/js/distillpub/template.v2.js"></script>
    <script src="/assets/js/distillpub/transforms.v2.js"></script>
    <script src="/assets/js/distillpub/overrides.js"></script>
    <style>
      p {
        text-align: justify;
        margin-bottom: 0;
      }
    </style>
    <style type="text/css">
      .fake-img {
        background: #bbb;
        border: 1px solid rgba(0, 0, 0, 0.1);
        box-shadow: 0 0 4px rgba(0, 0, 0, 0.1);
        margin-bottom: 12px
      }

      .fake-img p {
        font-family: monospace;
        color: white;
        text-align: left;
        margin: 12px 0;
        text-align: center;
        font-size: 16px
      }
    </style>
  </head>
  <body>
    <header>
      <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top">
        <div class="container">
          <a class="navbar-brand title font-weight-lighter" href="/">
            <span class="font-weight-bold">Workshop </span>SNP Calling on the VSC </a>
          <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar top-bar"></span>
            <span class="icon-bar middle-bar"></span>
            <span class="icon-bar bottom-bar"></span>
          </button>
          <div class="collapse navbar-collapse text-right" id="navbarNav">
            <ul class="navbar-nav ml-auto flex-nowrap">
              <li class="nav-item ">
                <a class="nav-link" href="/">about</a>
              </li>
              <li class="nav-item ">
                <a class="nav-link" href="/workshop/">workshop</a>
              </li>
              <li class="toggle-container">
                <button id="light-toggle" title="Change theme">
                  <i class="fa-solid fa-moon"></i>
                  <i class="fa-solid fa-sun"></i>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </nav>
      <progress id="progress" value="0">
        <div class="progress-container">
          <span class="progress-bar"></span>
        </div>
      </progress>
    </header>
    <div class="post distill">
      <d-title>
        <h1>Module 4: SNP calling - part 2</h1>
        <p>SNP cleaning and annotation</p>
      </d-title>
      <d-article>
        <d-contents>
          <nav class="l-text figcaption">
            <h3>Contents</h3>
            <div><a href="Clean the VCF file using vt">Clean the VCF file using vt</a></div>
          </nav>
        </d-contents>
        <style>
          p {
            text-align: justify;
          }
        </style>
      <h2 id="Clean the VCF file using vt">4.1 Clean the VCF file using vt</h2>
      <h2 id="SNP annotation with snpeff">4.1 SNP annotation with snpeff</h2>
      SNP annotation with snpeff
      <p>
        Run the vt commands to clean up the VCF file.
      </p>
      <div class="language-plaintext highlighter-rouge">
        <div class="highlight">
          <pre class="highlight"><code><span>cat raw_snps.vcf \</span>
<span>| vt decompose - \</span>
<span>| vt normalize -n -r hg38_chr22.fa - \</span>
<span>| vt uniq - \</span>
<span>| vt view -f "QUAL>30" -h - \</span>
<span>> clean_snps.vcf</span></code>
          </pre>
        </div>
      </div>
      <p>
        First, vt decompose turns multi-allelic variants into bi-allelic variants. According to the output, 11 variants are decomposed.
        Next, vt normalize converts the variants into a standardized format and vt uniq drops any duplicates.
        Finally, vt view filters out the SNPs with a QUAL value below your desired cutoff and the remaining SNPs are written to a file
        "clean_snps.vcf". 
      </p>
      <h2 id="SNP annotation with snpeff">4.2 SNP annotation with snpeff</h2>
      <p>
        Now that we have a list of cleaned SNPs, we can annotate them to predict their effect using snpeff.
        The prgram snpeff is distributed as a .jar file that can be run with java. First, define the location of the .jar file.
        If you followed the conda environment set-up, the file should be located in the share/snpeff-5.2-0 folder of your conda environment.
        We will also store the current directory in a DIR variable.
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>SNPEFF_JAR="${VSC_DATA}/miniconda3/envs/ws_snp_calling/share/snpeff-5.2-0/snpEff.jar"</span>
<span>DIR=$(pwd)</span></code>
            </pre>
          </div>
        </div> 
        Now, we can run snpeff, using:
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>java -jar ${SNPEFF_JAR} eff hg38 -dataDir $DIR \</span>
<span>clean_snps.vcf > annotated_snps.vcf</span></code>
            </pre>
          </div>
        </div> 
        The command instructs to run the snpeff file using java, annotate the SNPs using the hg38 genome (which will be downloaded) and store the output in 
        the current directory. 
        Have a look at the generated files.
        First, snpEff_summary.html contains statistics of the SNP annotation. Details for each SNP were written to
        annotated_snps.vcf. Finally, an overview of the affected genes are provided in the file snpEff_genes.txt. 
        <h3>4.2.1 Adapt snpEff setting to your organism</h3>
        Before running snpEff on another organism than H. sapiens, you should set the genome and genetic code table.
        Detailed instruction are provided at in the official documentation: 
        <a href="https://genomics.sschmeier.com/ngs-voi/#prepare-snpeff-database" target="_blank">https://genomics.sschmeier.com/ngs-voi/#prepare-snpeff-database</a> 
      </p>
      </d-article>
      <d-appendix>
        <d-citation-list>
        </d-citation-list>
      </d-appendix>
      <d-bibliography src="/assets/bibliography/2018-12-22-distill.bib"></d-bibliography>
    </div>
    <footer class="fixed-bottom">
      <div class="container mt-0"> Â© Copyright 2023 Sophie Colette. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. </div>
    </footer>
    <script src="/assets/js/bootstrap.bundle.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script>
    <script type="text/javascript">
      function progressBarSetup() {
        "max" in document.createElement("progress") ? (initializeProgressElement(), $(document).on("scroll", function() {
          progressBar.attr({
            value: getCurrentScrollPosition()
          })
        }), $(window).on("resize", initializeProgressElement)) : (resizeProgressBar(), $(document).on("scroll", resizeProgressBar), $(window).on("resize", resizeProgressBar))
      }

      function getCurrentScrollPosition() {
        return $(window).scrollTop()
      }

      function initializeProgressElement() {
        let e = $("#navbar").outerHeight(!0);
        $("body").css({
          "padding-top": e
        }), $("progress-container").css({
          "padding-top": e
        }), progressBar.css({
          top: e
        }), progressBar.attr({
          max: getDistanceToScroll(),
          value: getCurrentScrollPosition()
        })
      }

      function getDistanceToScroll() {
        return $(document).height() - $(window).height()
      }

      function resizeProgressBar() {
        progressBar.css({
          width: getWidthPercentage() + "%"
        })
      }

      function getWidthPercentage() {
        return getCurrentScrollPosition() / getDistanceToScroll() * 100
      }
      const progressBar = $("#progress");
      window.onload = function() {
        setTimeout(progressBarSetup, 50)
      };
    </script>
  </body>
</html>