<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title> Module 3 | SNP calling</title>
    <meta name="author" content="Sophie Colette">
    <meta name="description" content="Module 3: SNP calling - from fastq to vcf">
    <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website">
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04">
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous">
    <link defer rel="stylesheet" href="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.css">
    <link rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5">
    <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons">
    <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?19f3075a2d19613090fe9e16b564e1fe" media="" id="highlight_theme_light">
    <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20
																xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%E2%9A%9B%EF%B8%8F&lt;/text&gt;&lt;/svg&gt;">
    <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e">
    <link rel="canonical" href="https://sc-git1.github.io/blog/2021/distill/">
    <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?e74e74bf055e5729d44a7d031a5ca6a5" media="none" id="highlight_theme_dark">
    <script src="/assets/js/theme.js?96d6b3e1c3604aca8b6134c7afdd5db6"></script>
    <script src="/assets/js/dark_mode.js?9b17307bb950ffa2e34be0227f53558f"></script>
    <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.0.8/dist/medium-zoom.min.js" integrity="sha256-7PhEpEWEW0XXQ0k6kQrPKwuoIomz8R8IYyuU1Qew4P8=" crossorigin="anonymous"></script>
    <script defer src="/assets/js/zoom.js?7b30caa5023af4af8408a472dc4e1ebb"></script>
    <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script>
    <script type="text/javascript">
      window.MathJax = {
        tex: {
          tags: "ams"
        }
      };
    </script>
    <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script>
    <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
    <script src="/assets/js/distillpub/template.v2.js"></script>
    <script src="/assets/js/distillpub/transforms.v2.js"></script>
    <script src="/assets/js/distillpub/overrides.js"></script>
    <style>
      p {
        text-align: justify;
        margin-bottom: 0;
      }
    </style>
    <style type="text/css">
      .fake-img {
        background: #bbb;
        border: 1px solid rgba(0, 0, 0, 0.1);
        box-shadow: 0 0 4px rgba(0, 0, 0, 0.1);
        margin-bottom: 12px
      }

      .fake-img p {
        font-family: monospace;
        color: white;
        text-align: left;
        margin: 12px 0;
        text-align: center;
        font-size: 16px
      }
    </style>
  </head>
  <body>
    <header>
      <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top">
        <div class="container">
          <a class="navbar-brand title font-weight-lighter" href="/">
            <span class="font-weight-bold">Workshop </span>SNP Calling on the VSC </a>
          <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar top-bar"></span>
            <span class="icon-bar middle-bar"></span>
            <span class="icon-bar bottom-bar"></span>
          </button>
          <div class="collapse navbar-collapse text-right" id="navbarNav">
            <ul class="navbar-nav ml-auto flex-nowrap">
              <li class="nav-item ">
                <a class="nav-link" href="/">about</a>
              </li>
              <li class="nav-item ">
                <a class="nav-link" href="/workshop/">workshop</a>
              </li>
              <li class="toggle-container">
                <button id="light-toggle" title="Change theme">
                  <i class="fa-solid fa-moon"></i>
                  <i class="fa-solid fa-sun"></i>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </nav>
      <progress id="progress" value="0">
        <div class="progress-container">
          <span class="progress-bar"></span>
        </div>
      </progress>
    </header>
    <div class="post distill">
      <d-title>
        <h1>Module 3: SNP calling - part 1</h1>
        <p>From fastq to vcf</p>
      </d-title>
      <d-article>
        <d-contents>
          <nav class="l-text figcaption">
            <h3>Contents</h3>
            <div><a href="Data">Data</a></div>
            <div><a href="Create the Conda environment">Create the Conda environment</a></div>
            <div><a href="FastQC">FastQC</a></div>
            <div><a href="Mapping to the reference genome">Mapping to the reference genome</a></div>
            <div><a href="Quality control of the alignment">Quality control of the alignment</a></div>
          </nav>
        </d-contents>
        <style>
          p {
            text-align: justify;
          }
        </style>
        <h2 id="Data">Input data</h2>
        <p>In this workshop, we will run a SNP calling workflow on short read data produced on an Illumina GAIIx from exome-enriched DNA of chromosome 22 of a single human individual. THe data was generated as part of the 1000 genomes Genomes project. The dataset consists of one million 76bp reads. The data can be downloaded <a href="https://swift.rc.nectar.org.au:8888/v1/AUTH_a3929895f9e94089ad042c9900e1ee82/VariantDet_BASIC/NA12878.GAIIx.exome_chr22.1E6reads.76bp.fastq" target="_blank">here</a>. The reference genome of chromosome 22 of version h38 (latest major human genome release) can be downloaded at the bottom of this page: <a href="https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/" target="_blank">https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/</a>. Upload the files to the VSC, decompress the reference h38.fa by using the gunzip tool, and rename it to h38_22.fa to avoid confusion. </p>
        <h2 id="Create the Conda environment">Create the Conda environment</h2>
        <p>The following packages need to be installed: fastqc, bwa-mem2, samtools, bcftools, matplotlib, vt, and snpeff. First, create a new conda environment:
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>conda create --name ws_snp_calling</span></code></pre>
          </div>
        </div> 
        Next, activate the environment: 
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>conda activate ws_snp_calling</span></code></pre>
          </div>
        </div> 
        Finally, install the required packages: 
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>conda install bioconda::fastqc bioconda::bwa-mem2 bioconda::samtools bioconda::bcftools conda-forge::matplotlib bioconda::vt bioconda::snpeff</span></code>
				</pre>
          </div>
        </div> 
        Alternatively, you can add the Conda environment by downloading the <a href="/workshop/3_SNP_Calling_from_fastq_to_vcf/ws_snp_calling.yml" download>ws_snp_calling.yml file</a> file, uploading it to the VSC and running: 
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>conda env create --name ws_snp_calling -f ws_snp_calling.yml</span></code></pre>
          </div>
        </div> 
        Don't forget to activate the Conda environment: 
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>conda activate ws_snp_calling.yml</span></code></pre>
          </div>
        </div>
        </p>
        <h2 id="fastqc">fastqc</h2>
        <p>In this paragraph we will run a QC analysis on the fastq file. The first step is to inspect the file content. This is important as incomplete data transfer can happen when downloading a file. In a first step, we inspect the head of the file:
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>head NA12878.GAIIx.exome_chr22.1E6reads.76bp.fastq</span></code></pre>
          </div>
        </div>
        <figure>
          <picture>
            <source class="responsive-img-srcset" srcset=" //assets/img/head_fastq-480.webp 480w, /assets/img/head_fastq-800.webp 800w, /assets/img/head_fastq-1400.webp 1400w, " sizes="95vw" type="image/webp">
            </source>
            <img src="/assets/img/head_fastq.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
          </picture>
        </figure>
        <div class="caption">Output of head NA12878.GAIIx.exome_chr22.1E6reads.76bp.fastq</div> What does a file in the fastq format normally look like? Every read in a fastq file consists of 4 lines, also called 'fields'. </p>
        <ul>
          <li>
            <strong>Field 1</strong>: begins with the "@" character and contains the sequnce id, optionally followed by a description
          </li>
          <li>
            <strong>Field 2</strong>: the nucleotide sequence
          </li>
          <li>
            <strong>Field 3</strong>: begins with the "+" character, optionally follwed by the content of Field 1
          </li>
          <li>
            <strong>Field 4</strong>: contains encoded quality values for each position in the nucleotide sequence
          </li>
        </ul>
        <p>Similarly, the tail of the fastq file should end with Field 4 of the last read i.e., end with a complete read.</p>
        <figure>
          <picture>
            <source class="responsive-img-srcset" srcset=" //assets/img/head_fastq-480.webp 480w, /assets/img/head_fastq-800.webp 800w, /assets/img/head_fastq-1400.webp 1400w, " sizes="95vw" type="image/webp">
            </source>
            <img src="/assets/img/head_fastq.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
          </picture>
        </figure>
        <div class="caption">Output of tail NA12878.GAIIx.exome_chr22.1E6reads.76bp.fastq</div>
        <p>You can also check if the number of lines in the fastq file divided by 4 is equal to an integer.</p>
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span># Number of reads, should output: 1151702</span>
<span>wc -l NA12878.GAIIx.exome_chr22.1E6reads.76bp.fastq | awk '{x=$1/4; print x}'</span>
<span># Or equivalently, check if the remainder is equal to zero</span>
<span>wc -l NA12878.GAIIx.exome_chr22.1E6reads.76bp.fastq | awk '{x=$1%4; print x}'</span></code>
																									</pre>
          </div>
        </div>
        <p>This list is not exhaustive. You can come up with additional checks for your fastq file. For example, you can check whether the sequence in Field 2 and the quality scores in Field 4 of the same length. When you share a file or upload it to a server like the VSC, you can also check if the files before and after transfer are identical by generating checksums with a program like md5sum. If there are issues with the fastq file, try to download it again, or contact the provider of the original file to report the issue. </p>
        <p> After you controlled the validity of the fastq file, you can run the program fastqc to check the quality of the reads. For this, first create an directory for the output data:
        <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>mkdir -p fastqc_output</span></code></pre>
          </div>
        </div> and then run: <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>fastqc fastq_input_data/NA12878.GAIIx.exome_chr22.1E6reads.76bp.fastq -o fastqc_output</span></code>
				</pre>
          </div>
        </div>
        </p> Have a look at the list of files written to the output folder: <div class="language-plaintext highlighter-rouge">
          <div class="highlight">
            <pre class="highlight"><code><span>ls -lah ./fastqc_output</span></code></pre>
          </div>
        </div>
        <p>There should be two files in the output directory: a .zip file and a .html file. Download the .html file and open it in your browser. The .html file contains the FastQC report. The quality is measured by 10 statistics. The outcome is labeled in green (= pass), orange ("warning") or red ("error"). Images of the plots can be found in the .zip file, both in the PNG and SVG format. The FastQC report generated here contains 1 error and three warnings. However, note that a warning or an error does not necessarily mean that your data is of poor quality. Some statistics are expected to deviate based on the experiment you performed. It is therefore advised to always look up what the statistic means (within the context of your experiment) and the potential causes of the warning/error. In the next subparagraphs, we will go over all statistics and briefly explain how to interpret them.
         For more information we refer to the FastQC manual[TODO: REFERENCE!].
        </p>
        <h3>Basic statistics</h3>
        <figure>
          <picture>
            <source class="responsive-img-srcset" srcset=" //assets/img/stat1-480.webp 480w, /assets/img/stat1-800.webp 800w, /assets/img/stat1-1400.webp 1400w, " sizes="95vw" type="image/webp">
            </source>
            <img src="/assets/img/stat1.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
          </picture>
        </figure>
        <p>The basic statistics section never raises an error or warning, but informs you about some properties of the file:</p>
        <ul>
          <li>
            <strong>Filename</strong>
          </li>
          <li>
            <strong>File type</strong>
          </li>
          <li>
            <strong>Encoding</strong>: guess of the data format used, indicates the type of quality score (Field 4).
          </li> 
          <blockquote class="block-warning"> <h5 id="warning">WARNING</h5> 
            <p>
               The reported encoding is only a guess made by FastQC and can thus be wrong. 
               This happens primarly when the quality of your reads is overall of very good quality.
               Therefore, always check if the encoding is what you expected, as this has a large influence on the plot.
            </p>
          </blockquote>
          <p>
            The three possible quality score types are: Solexa+64, Phred+64 and Phred+33. 
            Some examples of data formats and their associated type are given in the table below.
          </p>
          <style>
            td,
            th {
              padding-top: 1%;
              padding-bottom: 1%;
              padding-left: 1%;
              padding-right: 1%;
            }

            td:nth-child(2) {
              text-align: justify;
              width: 60%;
            }

            td:nth-child(1) {
              width: 20%;
            }

            td:first-child,
            td:last-child {
              text-align: center;
            }
          </style>
          <table class="table" border=1px>
            <tr>
              <th>Encoding</th>
              <th>Quality score type</th>
            </tr>
            <tr>
              <td>Sanger / Illumina 1.8+</td>
              <td>Phred+33</td>
            </tr>
            <tr>
              <td>Sanger / Illumina 1.5</td>
              <td>Phred+64</td>
            </tr>
            <tr>
              <td>Sanger / Illumina 1.3</td>
              <td>Phred+64</td>
            </tr>
            <tr>
              <td>Solexa / Illumina</td>
              <td>Solexa+64</td>
            </tr>
          </table>
          <p>The type needs to be specified in tools like Trimmomatic (a read trimming tool for Illumina NGS data) and Bowtie2 (a reference alignment tool for short reads (50-100 bp)). Other tools only support one format, for example the reference algment tool "bwa-mem" only supports Phred+33 [reference:https://sourceforge.net/p/bio-bwa/mailman/bio-bwa-help/thread/CAF%2BL%2BTeKa9v0Nmjrn-DS9QKzsyTjAEcFyXMNkw-ygzLkdXo4AQ%40mail.gmail.com/#msg31652789]. Using Phred+64 files in a Phred+33 pipeline results in an incorrect interpretation of the quality score i.e., the quality is overestimated. Phred+33 is the standard in the newest Illumina versions, and the most widely supported format. To convert your fastq file to Phred+33, you can use for example the "FastQ Groomer" tool of Galaxy. </p>
          <li>
            <strong>Total sequences</strong>: Total number of sequences processed. It is good practice to keep track of the total number of reads throughout the analysis.
          </li>
          <li>
            <strong>Total bases</strong>: Total number of sequences processed.
          </li>
          <li>
            <strong>Sequences flagged as poor quality</strong>Sequences flagged as poor quality
          </li>
          <li>
            <strong>Sequence length</strong>: provides the length of the shortest and longest sequence. Returns one value if those are equal.
          </li>
          <li>
            <strong>%GC</strong>
          </li>
          <p>A question to ask yourself is: are the sequence length and %GC as expected?</p>
        </ul>
        <h3>Per base sequence quality</h3>
        <figure>
         <picture>
           <source class="responsive-img-srcset" srcset=" //assets/img/per_base_quality-480.webp 480w, /assets/img/per_base_quality-800.webp 800w, /assets/img/per_base_quality-1400.webp 1400w, " sizes="95vw" type="image/webp">
           </source>
           <img src="/assets/img/per_base_quality.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
         </picture>
        </figure>
        <p>The "per base sequence quality" plot is the most important analysis in the FastQC report. It shows the distribution of the quality score over all reads per position. The expected output 
            depends on the sequencing platform used. For example, in the case of sequencing-by-synthesis, the quality of calls degrade toward the end of the reads due to signal decay and phasing (= blurring of the signal due to loss of synchronicity within the clusters) towards the end of the run.
            The distribution of quality scores is represented by a whisker plot. The different aspects of the distribution and their visualisation are:
            <ul>
               <li><strong>Median</strong>: red line</li>
               <li><strong>Inter-quartile range (25-75%)</strong>: yellow box</li>
               <li><strong>10% threshold</strong>: lower whisker</li>
               <li><strong>90% threshold</strong>: upper whisker</li>
               <li><strong>Mean</strong>: blue line</li>
            </ul>
            The y-axis is split into 3 quality categories:
            <table class="table" border=1px>
               <tr>
                 <th>Quality</th>
                 <th>Color code</th>
               </tr>
               <tr>
                 <td>very good quality calls</td>
                 <td>green</td>
               </tr>
               <tr>
                 <td>reasonable quality calls</td>
                 <td>orange</td>
               </tr>
               <tr>
                  <td>poor quality</td>
                  <td>red</td>
                </tr>
             </table>
        </p>
        The quality scores are calculated using the following formula:
        $$-10 \log_{10}(e)$$ 
        <p>
         Thus, high scores indicate higher sequence quality. The error rate and base call accuracy of some scores are listed in the table below.
         <table class="table" border=1px>
            <tr>
              <th>Quality score</th>
              <th>Error rate</th>
              <th>Base call accuracy</th>
            </tr>
            <tr>
              <td>Q10</strong></td>
              <td>1 in 10</td>
              <td>90%</td>
            </tr>
            <tr>
               <td>Q20</strong></td>
               <td>1 in 100</td>
               <td>99%</td>
            </tr>
            <tr>
               <td>Q30</strong></td>
               <td>1 in 1000</td>
               <td>99.9%</td>
             </tr>
             <tr>
               <td>Q40</strong></td>
               <td>1 in 10,000</td>
               <td>99.99%</td>
             </tr>
          </table>
          Quality score above Q30 mean that virtually all of the reads will be perfect. 
          Q30 is considered a threshold for quality in next-generation sequencing (NGS).
        </p>
        <p>In the case of Illumina sequencing, frequenlty occurring probelms and their quality trend, are given in the table below.
            <table class="table" border=1px>
               <tr>
                 <th>Trend</th>
                 <th>Cause</th>
               </tr>
               <tr>
                 <td>Lower quality scores across the <strong>entire read</strong></td>
                 <td>Overclustering leading to overlapping signals</td>
               </tr>
               <tr>
                 <td>Sudden drops in signal or high % of poor quality reads</td>
                 <td>Instrumentation breakdown (e.g., cycles lost)</td>
               </tr>
             </table>
        </p>
        <blockquote class="block-tip"> <h5 id="Tip">TIP</h5> 
         <p>
            If the quality drops singificantly towards the end, you might consider trimming the reads by position, or based on quality.
         </p>
        </blockquote>
        <h3>Per tile sequence quality</h3>
        <p>The "per tile sequence quality" plot only appears in the presence of an Illumina library and the original sequence identifiers 
         (which contain the flowcell tile for each read). This plot shows the deviation from the average quality for each combination of flowcell tile (y-axis) and position (x-axis).
         The colors range from blue (at or above average) to red (below average). A good plot is entirely blue.
         An error is raised if any tile has a mean Phred score more than below the mean for that position across all tiles.<br>
         <strong>Interpretation</strong>
         <figure>
            <picture>
              <source class="responsive-img-srcset" srcset=" //assets/img/tile-480.webp 480w, /assets/img/tile-800.webp 800w, /assets/img/tile-1400.webp 1400w, " sizes="95vw" type="image/webp">
              </source>
              <img src="/assets/img/tile.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
            </picture>
           </figure>
         The error raised here is thus due to the variability in quality scores in tile 115 for a few positions. However, overall there do not appear to be 
         major problems with the calls made in tile 115. Note that FastQC's thresholds for raising warnings or errors are not absolute indicators of good or bad data. Instead, they are to help you identify potential quality issues.
         Therefore, we should interpret this error together with the other statistics.
        </p>
        <h3>Per sequence quality scores</h3>
        <figure>
         <picture>
           <source class="responsive-img-srcset" srcset=" //assets/img/per_sequence_quality-480.webp 480w, /assets/img/per_sequence_quality-800.webp 800w, /assets/img/per_sequence_quality-1400.webp 1400w, " sizes="95vw" type="image/webp">
           </source>
           <img src="/assets/img/per_sequence_quality.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
         </picture>
        </figure>
        The "Per sequence quality scores" plot shows the distribution of the average quality per read. Based on the figure, most reads have a high Phred score and thus are of very good quality.
        A peak below 27 (0.2% error rate) raises a warning, and a peak below 20 (= 1% error rate) raises an error.
        Moreover, a small local maximum may be visible at a low scores, which indicates that a subset of the sequences are of overall low quality.
        Often, the reason for some sequences being poorly imaged is their location at the edge of the field of view and thus this concerns only a small % of the sequences. 
        It is advised to quickly investigate the sequences of low quality, to identify the cause.
        <h3>Per base sequence content</h3>
        <figure>
         <picture>
           <source class="responsive-img-srcset" srcset=" //assets/img/per_base_sequence_content-480.webp 480w, /assets/img/per_base_sequence_content-800.webp 800w, /assets/img/per_base_sequence_content-1400.webp 1400w, " sizes="95vw" type="image/webp">
           </source>
           <img src="/assets/img/per_base_sequence_content.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
         </picture>
        </figure>
        <p>
         The "per base sequence content" plot shows the % of the 4 bases T, C, A and G per position. 
         In a random library, each position should have a base frequency that approximates the overall frequency in the genome.
         Deviations are the result of introduced bias during library preparation (e.g., overrepresented sequences) or during sequencing. 
         Some library preparation procedures will always introduce a technical bias, usually at the beginning of the reads.
         For example, the hexamer priming step of RNA-seq library preparation or libraries that result from fragmentation using transposases (= tagmentation), have enriched bases in the first 10-12 positions.
         This is usually not a problem for downstream analysis.
         A warning is raised when the difference between A/T or G/C in any position exceeds 10%. Errors are raised when the difference exceeds 20%.
        </p>
        <h3>Per sequence GC content</h3>
        <figure>
         <picture>
           <source class="responsive-img-srcset" srcset=" //assets/img/per_sequence_gc_content-480.webp 480w, /assets/img/per_sequence_gc_content-800.webp 800w, /assets/img/per_sequence_gc_content-1400.webp 1400w, " sizes="95vw" type="image/webp">
           </source>
           <img src="/assets/img/per_sequence_gc_content.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
         </picture>
        </figure>
        <p>
         The plot "per sequence GC content" usually contains a normal distribution with a mean equal to the GC content of the studied organism.
         The blue line represents the expected distribution based on the mean GC content. THe red line represents the actual distribution in the fastq file. Shifts of the normal distribution to the side due to systematic bias are not detected, since the program does not know the expected GC%.
         Sharp peaks on top of the normal distribution are an indication of contamination of overrepresentation sequences or a highly over-expressed gene. 
         A broad peak may indicate contamination with another organism.
         A warning/error is raised if the sum of the deviations from the normal distribution represents more than 15%/30% of the reads.
         If the total deviations from the standard distribution make up more than 15% or 30% of the reads, a warning or error are raised respectively.   
         </p>
        <h3>Per base N content</h3>
        <figure>
         <picture>
           <source class="responsive-img-srcset" srcset=" //assets/img/per_base_n_content-480.webp 480w, /assets/img/per_base_n_content-800.webp 800w, /assets/img/per_base_n_content-1400.webp 1400w, " sizes="95vw" type="image/webp">
           </source>
           <img src="/assets/img/per_base_n_content.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
         </picture>
        </figure>
        <p>
         The "per base N content" plot shows the % of N per position. An N is assigned each time that the confidence is not high enough to call a base. Thus, high percentages indicate that the data is not of sufficient quality to assure valid base calls. 
         A warning is raised if any position has an N content of more than 5%.
        </p>
        <h3>Sequence length distribution</h3>
        <figure>
         <picture>
           <source class="responsive-img-srcset" srcset=" //assets/img/sequence_length_distribution-480.webp 480w, /assets/img/sequence_length_distribution-800.webp 800w, /assets/img/sequence_length_distribution-1400.webp 1400w, " sizes="95vw" type="image/webp">
           </source>
           <img src="/assets/img/sequence_length_distribution.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
         </picture>
        </figure>
        <p>
         In our case, all sequences are of the same length. An error is raised if any sequence has a length equal to zero.
         A warning is raised when not all sequences are of the same length. 
         Based on the pipeline you used, the presence and degree of sequence variation are a metric of the quality of the base calls.
         For example, some pipelines remove poor quality base calls from the sequences.
        </p>
        <h3>Sequence duplication levels</h3>
        <figure>
         <picture>
           <source class="responsive-img-srcset" srcset=" //assets/img/duplication_levels.webp 480w, /assets/img/duplication_levels-800.webp 800w, /assets/img/duplication_levels-1400.webp 1400w, " sizes="95vw" type="image/webp">
           </source>
           <img src="/assets/img/duplication_levels.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
         </picture>
        </figure>
        <p>
         The sequence duplication levels are determined for the first 200,000 reads. Any read over 75bp is truncated to 50bp, because only exact matches count. 
         Because the likelihood of a sequencing error increases with read length, this procedure may lead to an underestimation of the duplication levels for longer sequences.
         Note that the upwards trend at > 10 in the figure is due to the binning.
        </p>
        <h3>Overrepresented sequences</h3>
        <p>
         There are no overrepresented sequences in the data. Overrepresented sequences are defined as sequences of  at least 20 bp with an occurrence of more than 0.1%.
         These are often remaining vector or adaptor sequences. They may explain skewed results in the "per sequence GC content" plot.
         <blockquote class="block-tip"> <h5 id="Tip">TIP</h5> 
            <p>
               If you don't recognize the overrepresented sequence immediately, you can identify it by using BLAST.
            </p>
           </blockquote>
        </p>
        <h3>Adapter content</h3>
        <figure>
         <picture>
           <source class="responsive-img-srcset" srcset=" //assets/img/adapter_content-480.webp 480w, /assets/img/adapter_content-800.webp 800w, /assets/img/adapter_content-1400.webp 1400w, " sizes="95vw" type="image/webp">
           </source>
           <img src="/assets/img/adapter_content.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" onerror="this.onerror=null; $('.responsive-img-srcset').remove();">
         </picture>
        </figure>
        <p>
         According to the figure, there is no overrepresentation of polyG, polyA, Nextera transposase, Illumina small RNA 5' adapter,
         Illumina small RNA 3' adapter or the Illumina universal adapter sequences.
        </p>
        <li>
         <h2 id="Mapping to the reference genome">Mapping to the reference genome</h2>
         <p>
            To call SNPs, the reads should first be mapped to the reference genome. In case you work with multiple conditions e.g., normal vs tumor samples, you should map all samples to the reference genome and compare the called SNPs to filter out the constitutional (germline) variants from the somatic (tumour) variants.
            Popular mapping tools are bwa-mem, bowtie2, and minimap2. Based on the literature, the tools perform similarly.
            In this workshop, we will use bwa-mem2, a speed optimized version of bwa-mem.
         </p>
         <p>
            First, create an index of the reference genome. An index makes looking up a seqeunce in the reference more efficient/faster.
            <div class="language-plaintext highlighter-rouge">
               <div class="highlight">
                 <pre class="highlight"><code><span>bwa-mem2 index h19_chr22.fa</span></code></pre>
               </div>
             </div>
            Then, align the reads and store the output in a SAM (= sequence aligment map) file:
            <div class="language-plaintext highlighter-rouge">
               <div class="highlight">
                 <pre class="highlight"><code><span>bwa-mem2 mem h19_chr22.fa NA12878.GAIIx.exome_chr22.1E6reads.76bp.fastq > NA12878.sam</span></code></pre>
               </div>
             </div>
            Sort the aligments based on their leftmost coordinates. By default, the output is is in the BAM (= binary aligment map) format. A bam file is the compressed (binary) representation of the sorted SAM file. 
            <div class="language-plaintext highlighter-rouge">
               <div class="highlight">
                 <pre class="highlight"><code><span>samtools sort NA12878.sam > NA12878.bam</span></code></pre>
               </div>
             </div>
             Finally, index the BAM file for faster data retrieval:
             <div class="language-plaintext highlighter-rouge">
               <div class="highlight">
                 <pre class="highlight"><code><span>samtools index NA12878.bam</span></code></pre>
               </div>
             </div>
             Have a look at the all the files that were created. Note the compression achieved with the BAM format (62M instead of 276M).
         </p>
         <h2 id="Quality control of the alignment">Quality control of the alignment</h2>
         <p>Now, let's have a look at the aligment results. You can display the header of the SAM file with:
            <div class="language-plaintext highlighter-rouge">
               <div class="highlight">
                 <pre class="highlight"><code><span>samtools view -H NA12878.sam</span></code></pre>
               </div>
             </div>
             The output should be:
             <div class="language-plaintext highlighter-rouge">
               <div class="highlight">
                 <pre class="highlight"><code><span>@SQ	SN:chr22	LN:51304566</span>
<span>@PG	ID:bwa-mem2	PN:bwa-mem2	VN:2.2.1	CL:bwa-mem2 mem h19_chr22.fa NA12878.GAIIx.exome_chr22.1E6reads.76bp.fastq</span>
<span>@PG	ID:samtools	PN:samtools	PP:bwa-mem2	VN:1.13	CL:samtools view -H NA12878.sam</span></code>
                 </pre>
               </div>
             </div>
             <h3>View header</h3>
               Header lines start with "@". The first line in the header contains information about the reference sequence i.e., the sequence name (SN) and the length (LN).
               The second line defines the program (PG) used to generate the SAM file:
               <ul>
                  <li><strong>ID</strong>: program identifier</li>
                  <li><strong>PN</strong>: program name</li>
                  <li><strong>VN</strong>: program version</li>
                  <li><strong>CL</strong>: the command run</li>
               </ul>
               Finally, the third line defines the program used to view the file (ID, PN, VN, CL). "PP" indicate the previous program run. 
             <h3>View alignments</h3>
             <p>
               Each line in a SAM file represents a record and consists of 11 mandatory fiels and optionally some extra fields. 
               As an example, the first record is:
               <div class="language-plaintext highlighter-rouge">
                  <div class="highlight">
                    <pre class="highlight"><code><span>61CC3AAXX100125:7:118:2538:5577	16	chr22	16050954	60	76M	*	0	0	GGGGATGAAGGCTCTTAGGTGGTAGCCCTCCAAGAGAATAGATGGTGAATGTCTCTTTTCAGACATTAAAGGTGTC	==?@AFIHEIIFFFHEIIIGIIEDIIIFFIIHIIIIHFEIIFFIIGIGFFHEEEEGGGDIHHEIDEBEEEFDAC>?	NM:i:0	MD:Z:76	AS:i:76	XS:i:0</span></code></pre>
                  </div>
                </div>
               The different columns are:
               <ul>
                  <li><strong>QNAME</strong>: name of the query sequence, <em>61CC3AAXX100125:7:118:2538:5577</em></li>
                  <li><strong>FLAG</strong>: representation of a 10-bit long sequence encoding the read's characteristics. For example, <em>16</em> stands for "mapped on the reverse strand"</li>
                  <li><strong>RNAME</strong>: name of the reference sequence, <em>chr22</em></li>
                  <li><strong>POS</strong>: 1-based leftmost position, <em>16050954</em></li>
                  <li><strong>MAPQ</strong>: mapping quality. In the case of bwa-mem, the quality ranges from 0 (= equally well mapping elsewhere) to 60. The value '255' indicates that the MAPQ value is not available.</li>
                  <li><strong>CIGAR</strong>: 1-based leftmost position, <em>16050954</em></li>
                  <li><strong>RNEXT</strong>: 1-based leftmost position, <em>16050954</em></li>
                  <li><strong>PNEXT</strong>: 1-based leftmost position, <em>16050954</em></li>
                  <li><strong>TLEN</strong>: 1-based leftmost position, <em>16050954</em></li>
                  <li><strong>SEQ</strong>: 1-based leftmost position, <em>16050954</em></li>
                  <li><strong>QUAL</strong>: 1-based leftmost position, <em>16050954</em></li>
               </ul>
             </p>

         </p>
          <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/workshop/5_project/">Module 5: SNP calling - part 2</a>
        </li>
      </d-article>
      <d-appendix>
        <d-citation-list>
         <p>https://emea.illumina.com/science/technology/next-generation-sequencing/plan-experiments/quality-scores.html</p>
         <p>https://training.galaxyproject.org/training-material/topics/sequence-analysis/tutorials/quality-control/tutorial.html</p>
        </d-citation-list>
      </d-appendix>
      <d-bibliography src="/assets/bibliography/2018-12-22-distill.bib"></d-bibliography>
    </div>
    <footer class="fixed-bottom">
      <div class="container mt-0"> Â© Copyright 2023 Sophie Colette. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. </div>
    </footer>
    <script src="/assets/js/bootstrap.bundle.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script>
    <script type="text/javascript">
      function progressBarSetup() {
        "max" in document.createElement("progress") ? (initializeProgressElement(), $(document).on("scroll", function() {
          progressBar.attr({
            value: getCurrentScrollPosition()
          })
        }), $(window).on("resize", initializeProgressElement)) : (resizeProgressBar(), $(document).on("scroll", resizeProgressBar), $(window).on("resize", resizeProgressBar))
      }

      function getCurrentScrollPosition() {
        return $(window).scrollTop()
      }

      function initializeProgressElement() {
        let e = $("#navbar").outerHeight(!0);
        $("body").css({
          "padding-top": e
        }), $("progress-container").css({
          "padding-top": e
        }), progressBar.css({
          top: e
        }), progressBar.attr({
          max: getDistanceToScroll(),
          value: getCurrentScrollPosition()
        })
      }

      function getDistanceToScroll() {
        return $(document).height() - $(window).height()
      }

      function resizeProgressBar() {
        progressBar.css({
          width: getWidthPercentage() + "%"
        })
      }

      function getWidthPercentage() {
        return getCurrentScrollPosition() / getDistanceToScroll() * 100
      }
      const progressBar = $("#progress");
      window.onload = function() {
        setTimeout(progressBarSetup, 50)
      };
    </script>
  </body>
</html>